go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -492,43 +570,54 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
-	klog.Info("Creating new shared gateway")
-	gw := &gateway{}
-
-	bridgeName, uplinkName, macAddress, ips, err := gatewayInitInternal(
-		nodeName, gwIntf, subnets, gwNextHops, nodeAnnotator)
-	if err != nil {
-		return nil, err
-	}
-
-	// the name of the patch port created by ovn-controller is of the form
-	// patch-<logical_port_name_of_localnet_port>-to-br-int
-	patchPort := "patch-" + bridgeName + "_" + nodeName + "-to-br-int"
-
-	// add masquerade subnet route to avoid zeroconf routes
+// addMasqueradeRoutes adds masquerade subnet route to avoid zeroconf routes
+func addMasqueradeRoutes(netIface string, gwNextHops []net.IP) error {
 	if config.IPv4Mode {
-		bridgeLink, err := util.LinkSetUp(bridgeName)
+		netIfaceLink, err := util.LinkSetUp(netIface)
 		if err != nil {
-			return nil, fmt.Errorf("unable to find shared gw bridge interface: %s", bridgeName)
+			return fmt.Errorf("unable to find interface: %s", netIface)
 		}
 		v4nextHops, err := util.MatchIPFamily(false, gwNextHops)
 		if err != nil {
-			return nil, fmt.Errorf("no valid ipv4 next hop exists: %v", err)
+			return fmt.Errorf("no valid ipv4 next hop exists: %v", err)
 		}
 		_, masqIPNet, _ := net.ParseCIDR(types.V4MasqueradeSubnet)
-		if exists, err := util.LinkRouteExists(bridgeLink, v4nextHops[0], masqIPNet); err == nil && !exists {
-			err = util.LinkRoutesAdd(bridgeLink, v4nextHops[0], []*net.IPNet{masqIPNet}, 0)
+		if exists, err := util.LinkRouteExists(netIfaceLink, v4nextHops[0], masqIPNet); err == nil && !exists {
+			err = util.LinkRoutesAdd(netIfaceLink, v4nextHops[0], []*net.IPNet{masqIPNet}, 0)
 			if err != nil {
 				if os.IsExist(err) {
 					klog.V(5).Infof("Ignoring error %s from 'route add %s via %s'",
 						err.Error(), masqIPNet, v4nextHops[0])
 				} else {
-					return nil, fmt.Errorf("unable to add OVN masquerade route to host, error: %v", err)
+					return fmt.Errorf("unable to add OVN masquerade route to host, error: %v", err)
 				}
 			}
 		} else if err != nil {
-			return nil, fmt.Errorf("failed to check if route exists for masquerade subnet, error: %v", err)
+			return fmt.Errorf("failed to check if route exists for masquerade subnet, error: %v", err)
+		}
+	}
+	return nil
+}
+
+func newSharedGateway(nodeName string, subnets []*net.IPNet, gwNextHops []net.IP, gwIntf string, gwIPs []*net.IPNet, nodeAnnotator kube.Annotator) (*gateway, error) {
+	klog.Info("Creating new shared gateway")
+	gw := &gateway{}
+
+	bridgeName, uplinkName, macAddress, ips, err := gatewayInitInternal(
+		nodeName, gwIntf, subnets, gwNextHops, gwIPs, nodeAnnotator)
+	if err != nil {
+		return nil, err
+	}
+
+	// the name of the patch port created by ovn-controller is of the form
+	// patch-<logical_port_name_of_localnet_port>-to-br-int
+	patchPort := "patch-" + bridgeName + "_" + nodeName + "-to-br-int"
+
+	if config.OvnKubeNode.Mode == types.NodeModeFull {
+		// add masquerade subnet route to avoid zeroconf routes
+		err = addMasqueradeRoutes(bridgeName, gwNextHops)
```

Comment:
dont you also need this on where you configure your service route for the smart nic host?